### PR TITLE
✨  introduce: Pluggable ActionEvaluator

### PR DESCRIPTION
**Pull request by riemannulus**
# Context
To ensure backward compatibility of `Action`, developing Lib9c Action was a challenging task. Therefore, to reduce development difficulty while implicitly maintaining backward compatibility, there was a need to change how `Action` is executed according to the height of the current `Block`. Therefore, we decided to name this project `Pluggable ActionEvaluator` and implement it.

# Rationale
[AssemblyLoadContext](https://learn.microsoft.com/en-us/dotnet/core/dependency-loading/understanding-assemblyloadcontext), for dynamic DLL loading, which we use for implementing `Puggable ActionEvaluator`.

We isolate `ActionEvaluator` as an `IPluginActionEvaluator` and use it on the `NineChronicles.Headless`. and get `IPluginKeyValueStore` for share `IKeyValueStore` with `NineChronicles.Headless`.

# Relative
- https://github.com/planetarium/lib9c/pull/2262
- https://github.com/planetarium/lib9c/commit/7ba6cb0050a539840dfb08e8bdf17576215f3671

### CODE REVIEW

File: .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs
Comment (moreal) [inline .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs:1]: What is a project to contain this file? It seems there is no `.Libplanet.Extensions.PluginActionEvaluator` project.

File: .Lib9c.Plugin/Lib9c.Plugin.csproj [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <EnableDynamicLoading>true</EnableDynamicLoading>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Libplanet.RocksDBStore" Version="3.8.1" />
```
Comment (moreal) [inline .Lib9c.Plugin/Lib9c.Plugin.csproj:11]: It seems unnecessary.

```suggestion
```

File: .Lib9c.Plugin/PluginActionEvaluator.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+using System.Security.Cryptography;
+using Lib9c.Plugin.Shared;
+using Libplanet.Action;
+using Libplanet.Common;
+using Libplanet.Extensions.ActionEvaluatorCommonComponents;
+using Libplanet.RocksDBStore;
```
Comment (moreal) [inline .Lib9c.Plugin/PluginActionEvaluator.cs:6]:
```suggestion
```

File: .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs
Comment (riemannulus) [reply .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs:1]: https://github.com/planetarium/NineChronicles.Headless/tree/feature/plugin-aev/Libplanet.Extensions.PluggedActionEvaluator Prepared on the `NineChronicles.Headless`. 😅

File: .Lib9c.Plugin/Lib9c.Plugin.csproj (hunk 1 above)
Comment (riemannulus) [reply .Lib9c.Plugin/Lib9c.Plugin.csproj]: https://github.com/planetarium/lib9c/pull/2265/commits/11807c7ffa13208ee9fa31495a3012d3adf8103a amended!

File: .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs
Comment (riemannulus) [reply .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs:1]: https://github.com/planetarium/lib9c/pull/2265/files#diff-0f35087f7996690ba003d3699f15b3c9dfddf9ebf22dde2ea1fe1806ef9b2e12

Aha, I'm misunderstanding. it is on the `Lib9c.Plugin`.

Comment (moreal) [reply .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs:1]: But it seems it doesn't include this file. (and not in the same directory)

https://github.com/planetarium/lib9c/pull/2265/files#diff-7e4500076379697b068d0f6f2e4d4d085e38eea2380974cdcc9773251ccf3325R10-R17

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <EnableDynamicLoading>true</EnableDynamicLoading>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\.Libplanet.Extensions.ActionEvaluatorCommonComponents\Libplanet.Extensions.ActionEvaluatorCommonComponents.csproj" />
    <ProjectReference Include="..\Lib9c\Lib9c.csproj" />
    <ProjectReference Include="..\.Lib9c.Plugin.Shared\Lib9c.Plugin.Shared.csproj">
      <Private>false</Private>
      <ExcludeAssets>runtime</ExcludeAssets>
    </ProjectReference>
  </ItemGroup>

</Project>
```

Comment (riemannulus) [reply .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs:1]: `Lib9c.Plugin.PluginActionEvaluator` is a wrapping class for making `ActionEvalautor` into `Plugin`. 

https://github.com/planetarium/lib9c/pull/2265/files#diff-0f35087f7996690ba003d3699f15b3c9dfddf9ebf22dde2ea1fe1806ef9b2e12R17-R24

It returns `byte[][]` when calling `Execute()` and it is not an `IActionEvaluator.Execute()` return type. So we need to wrap the class to convert the return type. It is `PluggedActionEvalautor` (https://github.com/planetarium/NineChronicles.Headless/tree/feature/plugin-aev/Libplanet.Extensions.PluggedActionEvaluator)

tl;dr: `PluginActionEvaluator` which executes the action. `PluggedActionEvalautor` is an adapter for `IActionEvalautor`.

Comment (riemannulus) [reply .Libplanet.Extensions.PluginActionEvaluator/PluginRocksDBKeyValueStore.cs:1]: Ok... we missed each... You mean `.Libplanet.Extensions.PluginActionEvaluator.PluginRocksDBKeyValueStore`, right? I missed that because I've already removed that project from my solution. I'll remove it. 👍
